### PR TITLE
feat(repo): Added prometheus and grafana & other implementations

### PR DESCRIPTION
**Pull request by 0xterminator**
- [x] added nats-prometheus explorer to infra
- [x] extended local docker-compose file with more services
- [x] added graceful shutdown to publisher + tests
- [x] added healtcheck handler to publisher + tests
- [x] added exposed metrics handler to publisher + tests
- [x]  Set up Prometheus server to collect metrics from our NATS servers and applications
- [x]  Configure NATS exporters to expose relevant metrics to Prometheus
- [x]  Set up Grafana server and connect it to the Prometheus data source
- [x]  Create Grafana dashboards for key metrics:
- [x]  NATS server health and performance
- [x]  Message throughput and latency
- [x]  Subscription counts and patterns
- [x]  Error rates and types
- [x]  Resource utilization (CPU, memory, network)
- [x]  Implement custom metrics in our application code where necessary
- [x]  Set up alerting rules in Prometheus and Grafana for critical thresholds
- [ ]  Create documentation for interpreting dashboards and responding to alerts
- [x]  Implement log aggregation and visualization in Grafana (consider using Loki)
- [x]  Set up proper authentication and authorization for Grafana access
- [ ]  Create a runbook for common operational tasks and troubleshooting scenarios

### CODE REVIEW

**Review by Jurshsmith — CHANGES_REQUESTED** [review]
Quite a large PR. Well done 👍🏾. 

I didn't go through it thoroughly, but I've left some comments on things I caught at a glance. Please let me know what you think.

File: crates/fuel-streams-core/src/stream/stream_impl.rs [hunk 1]
```diff
@@ -85,6 +85,7 @@ pub trait Streamable: StreamEncoder {
 /// TODO: Split this into two traits StreamPublisher + StreamSubscriber
 #[derive(Debug, Clone)]
 pub struct Stream<S: Streamable> {
+    client: NatsClient,
```
[the remark this answers is not in the document]
Comment (pedronauck) [reply crates/fuel-streams-core/src/stream/stream_impl.rs:88]: Agree

**Comment by pedronauck** [comment]
> Quite a large PR. Well done 👍🏾.
> 
> I didn't go through it thoroughly, but I've left some comments on things I caught at a glance. Please let me know what you think.

We should avoid merging this large and complex PR. It would be better to split it into smaller, more manageable PRs.

File: crates/fuel-streams-publisher/src/main.rs [hunk 2]
```diff
@@ -26,31 +46,63 @@ async fn main() -> anyhow::Result<()> {
...
         &cli.nats_url,
-        chain_id,
-        *base_asset_id,
-        fuel_core_database,
-        fuel_core_subscription,
+        state.metrics.clone(),
     )
     .await?;
-    publisher.run().await?;
+    tracing::info!("Publisher started.");
```
[the remark this answers is not in the document]
Comment (0xterminator) [reply crates/fuel-streams-publisher/src/main.rs]: agreed, changed

File: crates/fuel-streams-core/src/stream/stream_impl.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xterminator) [reply crates/fuel-streams-core/src/stream/stream_impl.rs]: agree also, changed.

File: crates/fuel-streams-core/src/stream/stream_impl.rs [hunk 3]
```diff
@@ -219,6 +221,17 @@ impl<S: Streamable> Stream<S> {
         }
     }
 
+    pub async fn flush_await(&self) -> Result<(), StreamError> {
```
[the remark this answers is not in the document]
Comment (0xterminator) [reply crates/fuel-streams-core/src/stream/stream_impl.rs]: done.

File: crates/fuel-streams-publisher/src/blocks.rs [hunk 4]
```diff
@@ -18,7 +27,72 @@ pub async fn publish(
...
+        .publishing_latency_histogram
+        .with_label_values(&[
+            &chain_id.to_string(),
+            &block_producer.to_string(),
+            BlocksSubject::WILDCARD,
+        ])
+        .observe(latency as f64);
+
+    metrics
+        .last_published_block_timestamp
```
[the remark this answers is not in the document]
Comment (0xterminator) [reply crates/fuel-streams-publisher/src/blocks.rs]: Not sure what you mean, you suggesting a macro here ?

File: docker/monitoring/grafana/provisioning/dashboards/nats-surveyor/surveyor-clients-dashboard.json
[the remark this answers is not in the document]
Comment (0xterminator) [reply docker/monitoring/grafana/provisioning/dashboards/nats-surveyor/surveyor-clients-dashboard.json:1]: there are files that go in these folders. I will add them to .gitignore however. thx

File: crates/fuel-streams-publisher/src/blocks.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xterminator) [reply crates/fuel-streams-publisher/src/blocks.rs]: I have resolved it differently, pls have a look

**Review by Jurshsmith — CHANGES_REQUESTED** [review]
Looks okay to merge. We just need to probably revert the `Box<dyn FuelCoreLike` change from your conflict resolution.

File: .gitignore [hunk 5]
```diff
@@ -15,3 +15,5 @@ docker/db
 tmp
 /docker/db
 build_rs_cov.profraw
+docker/monitoring/surveyor/jetstream
+docker/monitoring/surveyor/observations
```
Comment (Jurshsmith) [inline .gitignore:19]: Are these post-ignored? (it seems they are still part of your changes). It's fine if we want to hardcode them locally but I'm just making sure that we avoid git tracking them if they are meant to be auto-generated. what do you think?

File: crates/fuel-streams-publisher/src/publisher.rs [hunk 6]
```diff
@@ -40,28 +69,140 @@ impl Streams {
 }
 
 #[allow(dead_code)]
+/// TODO: Remove right after using chain_id and base_asset_id to publish
+/// TransactionsById subject
 pub struct Publisher {
-    streams: Streams,
-    fuel_core: Box<dyn FuelCoreLike>,
+    fuel_service: Arc<FuelService>,
+    fuel_core: Box<FuelCore>,
```
[the remark this answers is not in the document]
Comment (0xterminator) [reply crates/fuel-streams-publisher/src/publisher.rs]: I brought it back, was needed when refactoring the tests. fuel service is not inside fuel_core, fuel_core has the importer, chain id and hte db. We need an Arc of the service for the shutdown handlers and the share state.

**Comment by 0xterminator** [comment]
> Looks okay to merge. We just need to probably revert the `Box<dyn FuelCoreLike` change from your conflict resolution.

I re-introduced it already. Thanks.

File: .gitignore (hunk 5 above)
Comment (0xterminator) [reply .gitignore:19]: They are not autogenerated. They need to be there for the surveryor to work.

Comment (Jurshsmith) [reply .gitignore:19]: Then we shouldn't have them in the `.gitignore`, should incase we update their configurations (?) Sorry, I know I initially asked for it but it's now clear that they are regular Grafana dashboards

File: scripts/start-publisher.sh [hunk 7]
```diff
@@ -48,4 +48,5 @@ cargo run -p fuel-streams-publisher -- \
     --relayer-v2-listening-contracts $RELAYER_V2_LISTENING_CONTRACTS \
     --relayer-da-deploy-height $RELAYER_DA_DEPLOY_HEIGHT \
     --relayer-log-page-size $RELAYER_LOG_PAGE_SIZE \
-    --reserved-nodes /dns4/p2p-testnet.fuel.network/tcp/30333/p2p/16Uiu2HAmDxoChB7AheKNvCVpD4PHJwuDGn8rifMBEHmEynGHvHrf
+    --reserved-nodes /dns4/p2p-testnet.fuel.network/tcp/30333/p2p/16Uiu2HAmDxoChB7AheKNvCVpD4PHJwuDGn8rifMBEHmEynGHvHrf \
+    --server-addr 0.0.0.0:9000
```
Comment (Jurshsmith) [inline scripts/start-publisher.sh:52]: Do we want to update our `Dockerfile` as well?

File: .gitignore (hunk 5 above)
Comment (0xterminator) [reply .gitignore]: I have removed these 2 lines from gitignore now. In these folders we have configs that we should keep. Sorry about the confusion. All good now.

File: scripts/start-publisher.sh (hunk 7 above)
Comment (0xterminator) [reply scripts/start-publisher.sh:52]: We have to , yes. Now we are running an http server with metircs and healthchecks.

**Review by Jurshsmith — APPROVED** [review]

**Pull request merged**